### PR TITLE
Add a separate crate insn_decode in kernel for instruction decoding

### PR DESCRIPTION
**Pull request by cxdong**
This pull request introduces the kernel/insn_decode crate, which provides instruction decoding capabilities. It is implemented by referring the [ instruction decoder](https://github.com/projectacrn/acrn-hypervisor/blob/master/hypervisor/arch/x86/guest/instr_emul.c) from the ACRN project. As an acknowledgment, its copyright notice has been included in the header of decode.rs.

Upon a successful decoding for a given set of instruction bytes, a DecodedInsnCtx is created, containing detailed decoding outcomes. This structure exposes two public interfaces that allow access to the corresponding DecodedInsn and the decoded instruction length.

Currently, the crate supports decoding for the in/out, rdmsr, wrmsr, rdtsc, rdtscp, and cpuid instructions, which are essential for the current SEV #VC handler. So the #VC handler has been revised to utilize this crate for instruction decoding. The test suite is also expanded to cover all the aforementioned instructions. But due to the lack of an AMD machine on my end, this pull request has not been tested on an actual hardware. Appreciate it if someone can help to try this PR on an actual AMD machine.

### CODE REVIEW

**Review by 00xc — COMMENTED** [review]
Overall looks good, left some comments below. The nitpick is just a suggestion, you may ignore it.

**Comment by 00xc** [comment]
In-SVSM tests and regular boot work with these patches.

File: kernel/src/insn_decode/opcode.rs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    Out,
+    Rdmsr,
+    Rdtsc,
+    Rdtscp,
+    TwoByte,
+    Wrmsr,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct OpCodeDesc {
```
[the remark this answers is not in the document]
Comment (cxdong) [reply kernel/src/insn_decode/opcode.rs:45]: Sure, will add documents for all pub items.

File: kernel/src/insn_decode/insn.rs [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+        }
+
+        fn read_cr0(&self) -> u64 {
+            CR0Flags::PE.bits()
+        }
+
+        fn read_cr4(&self) -> u64 {
+            CR4Flags::LA57.bits()
+        }
+    }
```
[the remark this answers is not in the document]
Comment (cxdong) [reply kernel/src/insn_decode/insn.rs:126]: Sure, that is better!

File: kernel/src/insn_decode/insn.rs [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+
+    /// Decode the instruction with the given InsnMachineCtx.
+    ///
+    /// # Returns
+    ///
+    /// A [`DecodedInsnCtx`] if the instruction is supported, or an [`InsnError`] otherwise.
+    pub fn decode(&self, mctx: &dyn InsnMachineCtx) -> Result<DecodedInsnCtx, InsnError> {
+        DecodedInsnCtx::new(&self.0, mctx)
+    }
```
[the remark this answers is not in the document]
Comment (cxdong) [reply kernel/src/insn_decode/insn.rs:95]: Good cache! Indeed, it is not necessary to use dynamic dispatch and static dispatch is better.

File: kernel/src/insn_decode/decode.rs [hunk 4]
```diff
@@ -0,0 +1,746 @@
...
+    ///
+    /// # Returns
+    ///
+    /// An instruction byte if success or an [`InsnError`] otherwise
+    pub fn peek(&self) -> Result<u8, InsnError> {
+        if self.nr_processed < self.bytes.len() {
+            Ok(self.bytes[self.nr_processed])
+        } else {
+            Err(InsnError::InsnPeek)
+        }
```
[the remark this answers is not in the document]
Comment (cxdong) [reply kernel/src/insn_decode/decode.rs:78]: I like this rust style. Will update.

**Comment by cxdong** [comment]
Thanks @00xc for the help to try this PR on the hardware! Really appreciate that!

**Review by 00xc — APPROVED** [review]
LGTM. In the future we can even split this out into a separate package in the workspace, but for now this is good.

**Review by p4zuu — APPROVED** [review]
This looks good to me too. I don't find that the new decoder significantly increases the attack surface. The critical part is in the instruction emulation atm

**Pull request merged**